packages/solarwinds-apm/src/sampling/http.ts

```diff
@@ -0,0 +1,171 @@
+limitations under the License.
+*/
+
+import { context } from "@opentelemetry/api"
+import { suppressTracing } from "@opentelemetry/core"
+
+import { Backoff } from "../backoff.js"
+import { type Configuration } from "../config.js"
+import { componentLogger } from "../logger.js"
+import { Sampler } from "./sampler.js"
+
+const REQUEST_TIMEOUT = 10 * 1000 // 10s
+const RETRY_INITIAL_TIMEOUT = 500 // 500ms
+const RETRY_MAX_TIMEOUT = 60 * 1000 // 60s
+const RETRY_MAX_ATTEMPTS = 20
+const MULTIPLIER = 1.5
+
+/** Creates a timeout and unreferences it if running Node.js */
+function unrefTimeout(timeout: number, callback: () => void) {
+  const timer = setTimeout(callback, timeout)
+  if (typeof timer.unref === "function") {
+    timer.unref()
+  }
+  return timer
+}
+
+/** Retrieves the hostname (or User-Agent in browsers) in URL encoded format */
+export function hostname(): Promise<string> {
+  if (
+    typeof process !== "undefined" &&
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    typeof process?.versions?.node !== "undefined"
+  ) {
+    return import("node:os").then(({ hostname }) =>
+      encodeURIComponent(hostname()),
+    )
+  } else {
+    return Promise.resolve(encodeURIComponent(navigator.userAgent))
+  }
+}
+
+export class HttpSampler extends Sampler {
+  readonly #url: URL
+  readonly #headers: HeadersInit
+  readonly #service: string
+  readonly #hostname = hostname()
+  readonly #backoff = new Backoff({
+    initial: RETRY_INITIAL_TIMEOUT,
+    max: RETRY_MAX_TIMEOUT,
+    retries: RETRY_MAX_ATTEMPTS,
+    multiplier: MULTIPLIER,
+  })
+
+  #lastWarningMessage: string | undefined = undefined
+
+  constructor(config: Configuration) {
+    super(config, componentLogger(HttpSampler))
+
+    this.#url = config.collector
+    this.#service = encodeURIComponent(config.service)
+    this.#headers = config.headers
+
+    setImmediate(() => {
+      this.#loop()
+    })
+  }
+
+  override toString(): string {
+    return `HTTP Sampler (${this.#url.host})`
+  }
+
+  /** Logs a de-duplicated warning */
+  #warn(message: string, ...args: unknown[]) {
+    if (message !== this.#lastWarningMessage) {
+      this.logger.warn(message, ...args)
+      this.#lastWarningMessage = message
+    } else {
+      this.logger.debug(message, ...args)
+    }
+  }
+
+  /** Resets last memorised warning */
+  #resetWarn() {
+    this.#lastWarningMessage = undefined
+  }
+
+  /** Retries the request after an exponential backoff timeout */
+  #retry() {
+    const timeout = this.#backoff.backoff()
+    if (timeout) {
+      this.logger.debug(`retrying in ${(timeout / 1000).toFixed(1)}s`)
+      unrefTimeout(timeout, () => {
+        this.#loop()
+      })
+    } else {
+      this.logger.warn(
+        "Reached max retry attempts for sampling settings retrieval.",
+        "Tracing will remain disabled.",
+      )
+    }
+  }
+
+  /** Settings update loop */
+  #loop() {
```

Comment:
it looks like this would be a nice use case for web workers in browser?

Reply:
I do wish OTel actually offloaded most processing to a worker but as it is most of the code needs to run on the main worker and having just this loop run in a separate worker and communicate with the main one would add a lot of complexity for not much of a payoff.